src/Aspire.Hosting.Valkey/ValkeyResource.cs

```diff
@@ -0,0 +1,55 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+namespace Aspire.Hosting.ApplicationModel;
+
+/// <summary>
+/// A resource that represents a Valkey resource independent of the hosting model.
+/// </summary>
+/// <param name="name">The name of the resource.</param>
+public class ValkeyResource(string name) : ContainerResource(name), IResourceWithConnectionString
+{
+    internal const string PrimaryEndpointName = "tcp";
+
+    private EndpointReference? _primaryEndpoint;
+
+    /// <summary>
+    /// Gets the primary endpoint for the Valkey server.
+    /// </summary>
+    public EndpointReference PrimaryEndpoint => _primaryEndpoint ??= new(this, PrimaryEndpointName);
+
+    private ReferenceExpression ConnectionString =>
+        ReferenceExpression.Create(
+            $"{PrimaryEndpoint.Property(EndpointProperty.Host)}:{PrimaryEndpoint.Property(EndpointProperty.Port)}");
+
+    /// <summary>
+    /// Gets the connection string expression for the Valkey server.
+    /// </summary>
+    public ReferenceExpression ConnectionStringExpression
+    {
+        get
+        {
+            if (this.TryGetLastAnnotation<ConnectionStringRedirectAnnotation>(out var connectionStringAnnotation))
+            {
+                return connectionStringAnnotation.Resource.ConnectionStringExpression;
+            }
+
+            return ConnectionString;
+        }
+    }
+
+    /// <summary>
+    /// Gets the connection string for the Valkey server.
+    /// </summary>
+    /// <param name="cancellationToken"> A <see cref="CancellationToken"/> to observe while waiting for the task to complete.</param>
+    /// <returns>A connection string for the Valkey server in the form "host:port".</returns>
+    public ValueTask<string?> GetConnectionStringAsync(CancellationToken cancellationToken = default)
+    {
+        if (this.TryGetLastAnnotation<ConnectionStringRedirectAnnotation>(out var connectionStringAnnotation))
+        {
+            return connectionStringAnnotation.Resource.GetConnectionStringAsync(cancellationToken);
+        }
+
+        return ConnectionString.GetValueAsync(cancellationToken);
+    }
```

Comment:
```suggestion
    public ReferenceExpression ConnectionStringExpression => ConnectionString;
```

Reply:
Do we need to make this change to Redis as well? This was a copy paste.

Reply:
It only exists for redis because of the *PublishAsXXRedis pattern*.